### PR TITLE
qemu.tests: Add case for successful live snapshot transaction.

### PR DESCRIPTION
**Pull request by CongLi**
Signed-off-by: Cong Li <coli@redhat.com>

id: 1321839

Based on: https://github.com/autotest/tp-qemu/pull/701
https://github.com/avocado-framework/avocado-vt/pull/633

### CODE REVIEW

**Comment by CongLi** [comment]
17:11:29 DEBUG| Send command: {'execute': 'transaction', 'arguments': {'actions': [{'data': {'device': u'drive_image1', 'snapshot-file': '/usr/share/avocado/data/avocado-vt/images/image1-snap', 'mode': 'absolute-paths', 'format': 'qcow2'}, 'type': 'blockdev-snapshot-sync'}, {'data': {'device': u'drive_sn1', 'snapshot-file': '/usr/share/avocado/data/avocado-vt/images/sn1-snap', 'mode': 'absolute-paths', 'format': 'qcow2'}, 'type': 'blockdev-snapshot-sync'}, {'data': {'device': u'drive_sn2', 'snapshot-file': '/usr/share/avocado/data/avocado-vt/images/sn2-snap', 'mode': 'absolute-paths', 'format': 'qcow2'}, 'type': 'blockdev-snapshot-sync'}]}, 'id': 'Rv65KiGK'}
17:11:29 INFO | [qemu output] Formatting '/usr/share/avocado/data/avocado-vt/images/image1-snap', fmt=qcow2 size=21474836480 backing_file=/usr/share/avocado/data/avocado-vt/images/RHEL-Server-7.3-64-virtio.qcow2 backing_fmt=qcow2 encryption=off cluster_size=65536 lazy_refcounts=off refcount_bits=16
17:11:29 INFO | [qemu output] Formatting '/usr/share/avocado/data/avocado-vt/images/sn1-snap', fmt=qcow2 size=21474836480 backing_file=/usr/share/avocado/data/avocado-vt/images/sn1.qcow2 backing_fmt=qcow2 encryption=off cluster_size=65536 lazy_refcounts=off refcount_bits=16
17:11:29 INFO | [qemu output] Formatting '/usr/share/avocado/data/avocado-vt/images/sn2-snap', fmt=qcow2 size=21474836480 backing_file=/usr/share/avocado/data/avocado-vt/images/sn2.qcow2 backing_fmt=qcow2 encryption=off cluster_size=65536 lazy_refcounts=off refcount_bits=16

17:14:00 INFO | PASS 1-smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.live_snapshot.transaction

File: qemu/tests/live_snapshot_transaction.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            transaction_test.snapshot_file = image + "-snap"
+            snapshot_file = transaction_test.get_snapshot_file()
+            transaction_test.snapshot_args.update({"snapshot-file": snapshot_file})
+            args = {"type": "blockdev-snapshot-sync",
+                    "data": transaction_test.snapshot_args}
+            arg_list.append(args)
+
+        error_context.context("Create multiple live snapshots simultaneously"
+                              " with transaction", logging.info)
+        transaction_test.vm.monitor.transaction(arg_list)
```
Comment (vivianQizhu) [inline qemu/tests/live_snapshot_transaction.py]: We should check if transaction snapshots succeed or not here.

Comment (CongLi) [inline qemu/tests/live_snapshot_transaction.py]: Hi qizhu,

Updated, could you hel preview it again?

Thanks.

15:22:08 INFO | PASS 1-smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.live_snapshot.transaction

File: qemu/tests/live_snapshot_transaction.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    1. Boot up guest with a system disk and 2 data disk.
+    2. Create multiple live snapshots simultaneously for all 3 disks with transaction.
+    3. Check guest which should boot up and reboot successfully.
+
+    :param test: Kvm test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    image_chain = params.get("images")
```
Comment (xutian) [inline qemu/tests/live_snapshot_transaction.py]: image_chain = params.objects("images")

File: qemu/tests/live_snapshot_transaction.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            transaction_test.snapshot_args.update({"snapshot-file": snapshot_file})
+            args = {"type": "blockdev-snapshot-sync",
+                    "data": transaction_test.snapshot_args}
+            arg_list.append(args)
+
+        error.context("Create multiple live snapshots simultaneously"
+                      " with transaction", logging.info)
+        output = transaction_test.vm.monitor.transaction(arg_list)
+        # return nothing on successful transaction
+        if bool(output):
```
Comment (xutian) [inline qemu/tests/live_snapshot_transaction.py]: convert is unnecessary here.

Comment (CongLi) [inline qemu/tests/live_snapshot_transaction.py]: return is an empty dict "{}",
only "if output" will be failed.

File: qemu/tests/cfg/live_snapshot.cfg [hunk 4]
```diff
@@ -42,3 +42,15 @@
             dd_cmd = "dd if=/dev/zero of=%s bs=4K count=1024 oflag=dsync"
             dd_timeout = 600
             copy_file_timeout = 600
+        - transaction:
+            type = live_snapshot_transaction
+            images += " sn1 sn2"
+            image_name_sn1 = "images/sn1"
+            image_name_sn2 = "images/sn2"
+            snapshot_file = ${image_name} + "snap"
```
Comment (pingl2016) [inline qemu/tests/cfg/live_snapshot.cfg]: transaction_test.snapshot_file will be assigned with value "${images} + "-snap"" in the script, so I think the variable here is unnecessary.

**Comment by pingl2016** [comment]
Case passed after test it with avocado-framework/avocado-vt#633
smp_8.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.live_snapshot.transaction:  PASS (163.41 s)
 smp_8.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.live_snapshot.transaction:  PASS (115.61 s)

**Comment by pingl2016** [comment]
Acked-by: Ping Li pingl@redhat.com

File: qemu/tests/cfg/live_snapshot.cfg (hunk 4 above)
Comment (CongLi) [inline qemu/tests/cfg/live_snapshot.cfg]: Thanks @pingl2016 .
updated it. :)

**Comment by suqinhuang** [comment]
@CongLi conflicts error

**Comment by CongLi** [comment]
@suqinhuang

Conflicts solved, could you help review it?

Thanks.

File: qemu/tests/cfg/live_snapshot.cfg [hunk 5]
```diff
@@ -81,3 +81,14 @@
...
                     snapshot_mode = "existing"
                     match_str = "Could not open '%s': No such file or directory"
+        - transaction:
+            type = live_snapshot_transaction
+            images += " sn1 sn2"
+            image_name_sn1 = "images/sn1"
+            image_name_sn2 = "images/sn2"
+            check_base_image = yes
+            create_image = yes
+            create_image_image1 = no
```
Comment (suqinhuang) [inline qemu/tests/cfg/live_snapshot.cfg]: why need both create_image and force_create_image

File: qemu/tests/live_snapshot_transaction.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    1. Boot up guest with a system disk and 2 data disk.
+    2. Create multiple live snapshots simultaneously for all 3 disks with transaction.
+    3. Check guest which should boot up and reboot successfully.
+
+    :param test: Kvm test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    arg_list = []
+    try:
```
Comment (suqinhuang) [inline qemu/tests/live_snapshot_transaction.py]: according to the cases you wrote before, it's not good to use try, finally (without except)

Comment (CongLi) [inline qemu/tests/live_snapshot_transaction.py]: I think no need for this case, as there is exception handling.

Or where do you think it's a good place?

Comment (CongLi) [inline qemu/tests/live_snapshot_transaction.py]: @suqinhuang 

Could you help review it again?

Thanks.

**Comment by suqinhuang** [comment]
LGTM
Acked-by: Suqin Huang

**Comment by CongLi** [comment]
Thanks @pingl2016 @suqinhuang ,merged.

**Pull request merged**